pandas/core/generic.py

```diff
@@ -1126,44 +1144,40 @@ def rename(self, *args, **kwargs):
 
         See the :ref:`user guide <basics.rename>` for more.
         """
-        axes, kwargs = self._construct_axes_from_arguments(args, kwargs)
-        copy = kwargs.pop("copy", True)
-        inplace = kwargs.pop("inplace", False)
-        level = kwargs.pop("level", None)
-        axis = kwargs.pop("axis", None)
-        errors = kwargs.pop("errors", "ignore")
-        if axis is not None:
-            # Validate the axis
-            self._get_axis_number(axis)
-
-        if kwargs:
-            raise TypeError(
-                "rename() got an unexpected keyword "
-                'argument "{0}"'.format(list(kwargs.keys())[0])
-            )
-
-        if com.count_not_none(*axes.values()) == 0:
+        if mapper is None and index is None and columns is None:
             raise TypeError("must pass an index to rename")
 
-        self._consolidate_inplace()
+        if (index is not None or columns is not None):
+            if axis is not None:
+                raise TypeError("Cannot specify both 'axis' and any of 'index' or 'columns'")
+            elif mapper is not None:
+                raise TypeError("Cannot specify both 'mapper' and any of 'index' or 'columns'")
+        else:
+            # use the mapper argument
+            if axis in {1, "columns"}:
```

Comment:
We accept 3 arguments, `mapper`, `index` and `columns`. `mapper` can conflict with `index` or `columns` if either is provided, which the checks before this take care of. At this point if there aren't any conflicts, I think easiest to just combine the object

Note that this is arguably a little too strict, as this is no longer valid:

```python
df.rename({"key": "val"}, axis=1, index={"key2": "val2"})
```

But I don't think there is a lot to gain in supporting that given the variety of other options available here

Reply:
> Note that this is arguably a little too strict, as this is no longer valid:

I don't recall if that case was originally discussed, but I think it's OK to break it.